contrib/epee/include/serialization/wire/write.h

```diff
@@ -0,0 +1,287 @@
+    * `namespace wire { void write_bytes(writer&, const ns::type&); }`
+    * `namespace ns { void write_bytes(wire::writer&, const type&); }`
+
+  See `wrappers.h` for `is_array` requirements, and `traits.h` for `is_blob`
+  requirements. `write_bytes` function can also specify derived type for faster
+  output (i.e.  `namespace ns { void write_bytes(wire::epee_writer&, type&); }`).
+  Using the derived type allows the compiler to de-virtualize and allows for
+  custom functions not defined by base interface. Using base interface allows
+  for multiple formats with minimal instruction count. */
+
+namespace wire
+{
+  //! Interface for converting C/C++ objects to "wire" (byte) formats.
+  struct writer
+  {
+    writer() = default;
+
+    virtual ~writer() noexcept;
+
+    //! By default, insist on retrieving array size before writing array
+    static constexpr std::true_type need_array_size() noexcept { return{}; }
+
+    virtual void boolean(bool) = 0;
+
+    virtual void integer(std::intmax_t) = 0;
+    virtual void unsigned_integer(std::uintmax_t) = 0;
+
+    virtual void real(double) = 0;
+
+    virtual void string(boost::string_ref) = 0;
+    virtual void binary(epee::span<const std::uint8_t>) = 0;
+
+    virtual void start_array(std::size_t) = 0;
+    virtual void end_array() = 0;
+
+    virtual void start_object(std::size_t) = 0;
+    virtual void key(boost::string_ref) = 0;
+    virtual void binary_key(epee::span<const std::uint8_t>) = 0;
+    virtual void end_object() = 0;
+
+  protected:
+    writer(const writer&) = default;
+    writer(writer&&) = default;
+    writer& operator=(const writer&) = default;
+    writer& operator=(writer&&) = default;
+  };
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const bool source)
+  { dest.boolean(source); }
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const int source)
+  { dest.integer(source); }
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const long source)
+  { dest.integer(std::intmax_t(source)); }
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const long long source)
+  { dest.integer(std::intmax_t(source)); }
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const unsigned source)
+  { dest.unsigned_integer(source); }
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const unsigned long source)
+  { dest.unsigned_integer(std::uintmax_t(source)); }
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const unsigned long long source)
+  { dest.unsigned_integer(std::uintmax_t(source));}
+
+  template<typename W>
+  inline void write_arithmetic(W& dest, const double source)
+  { dest.real(source); }
+
+  // Template both arguments to allow derived writer specializations
+  template<typename W, typename T>
+  inline std::enable_if_t<std::is_arithmetic<T>::value> write_bytes(W& dest, const T source)
+  { write_arithmetic(dest, source); }
```

Comment:
The problem is `bool` gets accepted for one of those types. There's probably a SFINAE method to fix that too, but I didn't bother coming up with it.

Reply:
I misspoke, I did this in the read interface, not sure why I didn't do it here. So this will be updated, along with the namespace recommendation and some squashing.

Reply:
Nope, still getting the ambiguity errors, despite it being nearly identical to the read code. I'm not wasting further time on this, maybe in a future PR this can get cleaned up a bit.